File: android-client/app/src/main/java/com/funwithactivity/app/core/state/AppState.java
package com.funwithactivity.app.core.state;

import java.util.Collections;
import java.util.LinkedHashMap;
import java.util.List;
import java.util.Map;

import funwithactivity.recommendations.v1.Recommendations.ProviderStatus;

/**
 * Shared client state: the measurements and developer fault settings the
 * Profile tab edits, plus a dirty flag and the most recent fetch result.
 *
 * Mirrors web-client's {@code funwithactivity.app.AppState}
 * (web-client/src/features/app/app-state.js): same three concerns
 * (measurements, faults, a dirty flag), same touch-only-on-real-change
 * discipline. The dirty flag exists so the Recommendations tab can refetch
 * on becoming visible ONLY when something actually changed — refetching
 * unconditionally burns a vendor call on every tab switch (the vendors are
 * flaky and rate-limited); never refetching makes the app look like it
 * ignored the user. See the Recommendations fragment for the consumer side
 * of this contract.
 *
 * One process-lifetime instance, held by {@link
 * com.funwithactivity.app.FunWithActivityApplication}. Not persisted to
 * disk and not thread-confined to the UI thread — all mutators are
 * synchronized because the gRPC fetch happens on a background executor
 * while Profile edits happen on the UI thread.
 */
public final class AppState {

    // Demo-friendly defaults so the app is populated with results on first
    // launch without requiring a Profile visit first: both providers engage
    // (service2 requires a birth date; service1 does not). Clearing the
    // birth date from Profile is the data-minimisation demo beat — it drops
    // service2 out of the response and shows the skipped banner.
    private static final double DEFAULT_HEIGHT_CM = 175.0;
    private static final double DEFAULT_WEIGHT_KG = 70.0;
    /** 1990-01-01T00:00:00Z, matches the demo constant used elsewhere in this client. */
    private static final long DEFAULT_BIRTH_DATE_UNIX = 631152000L;

    private double heightCm = DEFAULT_HEIGHT_CM;
    private double weightKg = DEFAULT_WEIGHT_KG;
    /** 0 means "not supplied" — mirrors the proto's own sentinel. */
    private long birthDateUnix = DEFAULT_BIRTH_DATE_UNIX;

    /** Provider name -> fault mode ("error" | "timeout" | "malformed"). */
    private final Map<String, String> faults = new LinkedHashMap<>();

    private boolean dirty = false;

    /**
     * Set true the instant the first fetch completes (success or failure).
     * The Recommendations tab fetches once on first appearance regardless
     * of {@link #dirty}; after that, only a dirty tab triggers a refetch.
     */
    private boolean everFetched = false;

    /** Statuses from the most recently completed fetch; feeds the Sources tab. */
    private List<ProviderStatus> lastStatuses = Collections.emptyList();

    public synchronized double getHeightCm() {
        return heightCm;
    }

    public synchronized double getWeightKg() {
        return weightKg;
    }

    public synchronized long getBirthDateUnix() {
        return birthDateUnix;
    }

    /**
     * Writes measurements, marking state dirty only if a value actually
     * changed — an unchanged write (e.g. re-focusing and blurring a field
     * without editing it) must not schedule a refetch.
     */
    public synchronized void setMeasurements(double heightCm, double weightKg, long birthDateUnix) {
        if (this.heightCm == heightCm && this.weightKg == weightKg && this.birthDateUnix == birthDateUnix) {
            return;
        }
        this.heightCm = heightCm;
        this.weightKg = weightKg;
        this.birthDateUnix = birthDateUnix;
        touch();
    }

    /** Defensive copy — callers must not mutate the live fault map. */
    public synchronized Map<String, String> getFaults() {
        return new LinkedHashMap<>(faults);
    }

    public synchronized void setFault(String provider, String mode) {
        if (mode.equals(faults.get(provider))) {
            return;
        }
        faults.put(provider, mode);
        touch();
    }

    public synchronized void clearFault(String provider) {
        if (!faults.containsKey(provider)) {
            return;
        }
        faults.remove(provider);
        touch();
    }

    public synchronized boolean isDirty() {
        return dirty;
    }

    /** Called by the Recommendations tab once it has fetched against current state. */
    public synchronized void markClean() {
        dirty = false;
    }

    public synchronized boolean hasFetchedOnce() {
        return everFetched;
    }

    public synchronized void setFetchedOnce() {
        everFetched = true;
    }

    public synchronized List<ProviderStatus> getLastStatuses() {
        return lastStatuses;
    }

    public synchronized void setLastStatuses(List<ProviderStatus> statuses) {
        this.lastStatuses = statuses;
    }

    private void touch() {
        dirty = true;
    }
}

File: web-client/src/features/app/app-state.js
goog.provide('funwithactivity.app.AppState');

goog.require('goog.events.EventTarget');


/**
 * Shared client state: the measurements the Profile screen edits and the
 * fault settings its DEVELOPER section toggles.
 *
 * The dirty flag exists so the Recs screen can refetch on becoming visible
 * ONLY when something actually changed. Refetching unconditionally would
 * spend a vendor call on every tab switch, and these vendors are flaky
 * cold-start Lambdas; never refetching would make the demo silently lie —
 * the presenter changes a value, returns, and the screen does not move.
 * @constructor
 * @extends {goog.events.EventTarget}
 */
funwithactivity.app.AppState = function() {
  funwithactivity.app.AppState.base(this, 'constructor');

  /**
   * Height and weight start at plausible defaults — matching iOS
   * (FWAAppState.m) and Android (AppState.java) — so the app opens on real
   * recommendations rather than an empty state. Birth date deliberately
   * does NOT: see below.
   * @private {number}
   */
  this.heightCm_ = funwithactivity.app.AppState.DEFAULT_HEIGHT_CM;
  /** @private {number} */
  this.weightKg_ = funwithactivity.app.AppState.DEFAULT_WEIGHT_KG;
  /**
   * Unix seconds, or 0 for "not supplied" — a deliberate GDPR Art. 5(1)(c)
   * data-minimisation choice, matching the wire convention web-proxy uses.
   *
   * Unset by default, and that default is load-bearing rather than
   * incidental: the app opens with service2 already skipped, so supplying
   * an age visibly UNLOCKS a provider instead of withholding one. Stating
   * the minimisation rule that way round — you get more by sharing more,
   * and the product works either way — reads as a product decision rather
   * than as a degraded mode.
   * @private {number}
   */
  this.birthDateUnix_ = 0;
  /** @private {!Object<string,string>} */
  this.faults_ = {};
  /** @private {boolean} */
  this.dirty_ = false;
};
goog.inherits(funwithactivity.app.AppState, goog.events.EventTarget);


/**
 * Default measurements, kept identical to iOS FWAAppState.m and Android
 * AppState.java so all three clients open on the same result set.
 * @const {number}
 */
funwithactivity.app.AppState.DEFAULT_HEIGHT_CM = 175;


/** @const {number} */
funwithactivity.app.AppState.DEFAULT_WEIGHT_KG = 70;


/**
 * Where the birth-date picker opens when nothing is set. The field itself
 * stays unset — this only positions the wheel somewhere plausible instead
 * of at an arbitrary epoch, so picking a date on stage is one gesture
 * rather than thirty years of scrolling.
 * @const {string}
 */
funwithactivity.app.AppState.PICKER_DEFAULT_DATE = '1983-05-29';


/** @const {string} */
funwithactivity.app.AppState.CHANGE = 'app-state-change';


/** @return {{heightCm: number, weightKg: number, birthDateUnix: number}} */
funwithactivity.app.AppState.prototype.getMeasurements = function() {
  return {
    heightCm: this.heightCm_,
    weightKg: this.weightKg_,
    birthDateUnix: this.birthDateUnix_
  };
};


/**
 * Writes measurements, marking state dirty only if a value actually
 * changed. An unchanged write must not schedule a refetch — otherwise
 * merely visiting Profile and leaving would burn a vendor call.
 * @param {number} heightCm
 * @param {number} weightKg
 * @param {number} birthDateUnix
 */
funwithactivity.app.AppState.prototype.setMeasurements = function(
    heightCm, weightKg, birthDateUnix) {
  if (this.heightCm_ === heightCm && this.weightKg_ === weightKg &&
      this.birthDateUnix_ === birthDateUnix) {
    return;
  }
  this.heightCm_ = heightCm;
  this.weightKg_ = weightKg;
  this.birthDateUnix_ = birthDateUnix;
  this.touch_();
};


/** @return {!Object<string,string>} */
funwithactivity.app.AppState.prototype.getFaults = function() {
  const copy = {};
  for (const k in this.faults_) copy[k] = this.faults_[k];
  return copy;
};


/**
 * @param {string} provider
 * @param {string} mode
 */
funwithactivity.app.AppState.prototype.setFault = function(provider, mode) {
  if (this.faults_[provider] === mode) return;
  this.faults_[provider] = mode;
  this.touch_();
};


/** @param {string} provider */
funwithactivity.app.AppState.prototype.clearFault = function(provider) {
  if (!(provider in this.faults_)) return;
  delete this.faults_[provider];
  this.touch_();
};


/** @return {boolean} */
funwithactivity.app.AppState.prototype.isDirty = function() {
  return this.dirty_;
};


/** Called by the Recs screen once it has fetched against current state. */
funwithactivity.app.AppState.prototype.markClean = function() {
  this.dirty_ = false;
};


/** @private */
funwithactivity.app.AppState.prototype.touch_ = function() {
  this.dirty_ = true;
  this.dispatchEvent(funwithactivity.app.AppState.CHANGE);
};
